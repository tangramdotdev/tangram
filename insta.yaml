test:
  auto_review: true
  runner: nextest
