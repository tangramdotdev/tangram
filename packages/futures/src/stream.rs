use self::attach::Attach;
use self::take_while_inclusive::TakeWhileInclusive;
use futures::{Future, Stream, StreamExt as _, TryStream, TryStreamExt as _};

pub mod attach;
pub mod take_while_inclusive;

pub trait Ext: Stream {
	fn attach<T>(self, value: T) -> Attach<Self, T>
	where
		Self: Sized,
	{
		Attach::new(self, value)
	}

	fn last(mut self) -> impl Future<Output = Option<Self::Item>>
	where
		Self: Sized + Unpin,
	{
		async move {
			let mut last = None;
			while let Some(item) = self.next().await {
				last = Some(item);
			}
			last
		}
	}

	fn take_while_inclusive<F, Fut>(self, predicate: F) -> TakeWhileInclusive<Self, F, Fut>
	where
		Self: Sized,
		F: FnMut(&Self::Item) -> Fut,
		Fut: Future<Output = bool>,
	{
		TakeWhileInclusive::new(self, predicate)
	}
}

impl<S> Ext for S where S: Stream {}

pub trait TryExt: TryStream {
	fn try_last(mut self) -> impl Future<Output = Result<Option<Self::Ok>, Self::Error>>
	where
		Self: Sized + Unpin,
	{
		async move {
			let mut last = None;
			while let Some(item) = self.try_next().await? {
				last = Some(item);
			}
			Ok(last)
		}
	}
}

impl<S> TryExt for S where S: TryStream {}
