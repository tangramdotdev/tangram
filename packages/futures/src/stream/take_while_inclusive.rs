use futures::{Future, Stream};
use pin_project::pin_project;
use std::{
	pin::Pin,
	task::{Context, Poll},
};

#[pin_project]
pub struct TakeWhileInclusive<S, F, Fut>
where
	S: Stream,
{
	#[pin]
	stream: S,
	predicate: F,
	#[pin]
	pending_predicate: Option<Fut>,
	pending_item: Option<S::Item>,
	done: bool,
}

impl<S, F, Fut> TakeWhileInclusive<S, F, Fut>
where
	S: Stream,
	F: FnMut(&S::Item) -> Fut,
	Fut: Future<Output = bool>,
{
	pub fn new(stream: S, predicate: F) -> Self {
		Self {
			stream,
			predicate,
			pending_predicate: None,
			pending_item: None,
			done: false,
		}
	}
}

impl<S, F, Fut> Stream for TakeWhileInclusive<S, F, Fut>
where
	S: Stream,
	F: FnMut(&S::Item) -> Fut,
	Fut: Future<Output = bool>,
{
	type Item = S::Item;

	fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
		let mut this = self.project();

		if *this.done {
			return Poll::Ready(None);
		}

		loop {
			if let Some(pending) = this.pending_predicate.as_mut().as_pin_mut() {
				match pending.poll(cx) {
					Poll::Ready(true) => {
						let item = this.pending_item.take().unwrap();
						this.pending_predicate.set(None);
						return Poll::Ready(Some(item));
					},
					Poll::Ready(false) => {
						let item = this.pending_item.take().unwrap();
						*this.done = true;
						return Poll::Ready(Some(item));
					},
					Poll::Pending => return Poll::Pending,
				}
			}

			match this.stream.as_mut().poll_next(cx) {
				Poll::Ready(Some(item)) => {
					let future = (this.predicate)(&item);
					*this.pending_item = Some(item);
					this.pending_predicate.set(Some(future));
				},
				Poll::Ready(None) => {
					*this.done = true;
					return Poll::Ready(None);
				},
				Poll::Pending => return Poll::Pending,
			}
		}
	}
}

#[cfg(test)]
mod tests {
	use crate::stream::Ext as _;
	use futures::{future, stream, StreamExt as _};

	#[tokio::test]
	async fn take_while_inclusive() {
		let values = stream::iter([1, 2, 3, 4, 5])
			.take_while_inclusive(|value| future::ready(*value != 4))
			.collect::<Vec<_>>()
			.await;
		assert_eq!(values, [1, 2, 3, 4]);
	}
}
