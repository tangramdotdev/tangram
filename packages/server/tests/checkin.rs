use insta::assert_yaml_snapshot;
use std::pin::pin;
use tangram_client as tg;
use tangram_futures::stream::TryStreamExt as _;
use tangram_server::{Config, Server};
use tangram_temp::{artifact, Temp};

type Result = std::result::Result<(), Box<dyn std::error::Error>>;

#[tokio::test]
async fn test() -> Result {
	let temp = Temp::new();
	let options = Config::with_path(temp.path().to_owned());
	let server = Server::start(options).await?;
	let directory = Temp::new();
	artifact!({
		"directory": {
			"README.md": "Hello, World!",
		}
	})
	.to_path(&directory)
	.await?;
	let arg = tg::artifact::checkin::Arg {
		destructive: false,
		deterministic: false,
		ignore: true,
		locked: false,
		path: directory.as_ref().join("package"),
	};
	let stream = server.check_in_artifact(arg).await?;
	let output = pin!(stream)
		.try_last()
		.await?
		.and_then(|event| event.try_unwrap_output().ok())
		.ok_or_else(|| tg::error!("stream ended without output"))?;
	assert_yaml_snapshot!(output);
	Ok(())
}
