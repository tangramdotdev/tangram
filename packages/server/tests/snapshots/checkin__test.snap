---
source: packages/server/tests/checkin.rs
expression: output
---
artifact: dir_01419kh5mw2t22j1gdcmxszbwn5npr5886fg5d412w92a0v4a1g5fg
